### ISSUE TITLE
Validation output diverges to very large constants during GM12878 Hi-C/OE fine-tuning, while training loss remains low

### ISSUE THREAD

**Issue by liushiji0511** [liu-bioinfo-lab/EPCOT#9]
Hi, thank you for releasing this code.

I am trying to reproduce the Hi-C/OE fine-tuning pipeline on GM12878, but I am encountering severe training/validation mismatch and unstable validation behavior. I would like to ask whether I may be using the code incorrectly, or whether there are some implementation details or preprocessing steps that are not explicit in the repository.

1. Main issue

During training, the training loss looks relatively small and stable, but the validation output quickly collapses to very large constant-like values, and the validation loss explodes.

For example, after clipping my validation labels to [-2, 2], the labels look numerically reasonable, but the model predictions still drift to very large negative values. In my logs, the validation labels are typically in [-2, 2] with means close to zero, while the validation outputs quickly move to values such as -70, then -1300, then -3000, and even lower in later epochs, with validation loss growing from thousands to millions. Pearson and Spearman correlations also become nan because the predictions appear close to constant outputs.

2. What I have already checked
(a) Distance embedding indices

I checked the indices passed to distance_embed, and they seem valid:

pos shape = (1, 200, 200)
dtype = int64
min = 0, max = 10
embedding num = 11

So this does not look like an embedding index out-of-range problem.

(b) Label range

My target values are already clipped / normalized to a small range, and the validation labels look numerically normal, but the model outputs still diverge badly. The issue therefore seems to be on the model side rather than simply due to extreme label values.

(c) Environment compatibility

I initially had a NumPy/SciPy version warning, but after fixing the environment that warning disappeared. The instability issue remains, so it does not seem to be caused by that.

3. Suspected source of instability in the model head

After reading the model implementation, I suspect the pairwise output head may be contributing to the instability:

def output_head(self,x,dist_embed,bins):
    x1=torch.tile(x.unsqueeze(1),(1,bins,1,1))
    x2 = x1.permute(0,2,1,3)
    mean_out=(x1+x2)/2
    dot_out=x1*x2
    return mean_out+dot_out+dist_embed

In particular, the term dot_out = x1 * x2 is an elementwise product, and it is added directly to mean_out and dist_embed. The final output is then produced by the dilated tower and a linear layer, without any obvious final output-range constraint.

This makes me wonder:

Is dot_out = x1 * x2 the intended design here?
When training on OE / clipped OE targets, is there an additional target transformation that should be applied, such as log1p, asinh, z-score normalization, or something similar?
Should the final prediction head have an output constraint or activation, depending on the target range?
What target preprocessing was used in your original experiments?
4. Scheduler / gradient accumulation question

I also noticed that the training script uses gradient accumulation (accum_iter=2), but scheduler.step() is called every batch, while optimizer.step() is called only when accumulation finishes:

loss = criterion(output, input_label)
loss = loss / args.accum_iter
loss.backward()
if ((step + 1) % args.accum_iter == 0) or (step + 1 == len(train_loader)):
    optimizer.step()
    optimizer.zero_grad()

scheduler.step()

This means the OneCycleLR schedule is not synchronized with the actual parameter updates.

Was this intentional, or should scheduler.step() be called only after optimizer.step()?

5. Minor issue: model saving directory

The script saves checkpoints to:

torch.save(model.state_dict(), 'models/%s_%s.pt' % (cl, args.bins))

If the models/ directory does not already exist, this raises an error.

This is easy to fix locally, but I wanted to mention it in case the code assumes that directory already exists.

6. My main questions

I would especially appreciate clarification on the following points:

Does this kind of behavior — relatively small training loss but validation predictions collapsing to very large constant-like values — usually indicate incorrect target preprocessing?
In your original setup, did you apply any transformation to the Hi-C/OE target before using MSE?
Is the pairwise head with dot_out = x1 * x2 expected to remain numerically stable as-is, or is it meant to be used together with additional normalization/scaling?
Should OneCycleLR be stepped only when optimizer.step() is called?

If you have a recommended training configuration for reproducing the original results, especially for:

target preprocessing
learning rate
scheduler usage
clipping / normalization
expected checkpoint directory setup

that would be extremely helpful.

Thank you very much.

If you want, I can also make this into a more compact GitHub-issue style version.

**Comment by linghuajiang** [comment]
Hi, thank you for your interest in EPCOT. I think the main problem might be your target data preparation. In our original pipeline, Hi-C targets are raw OE values with no additional transformation – no log, no clipping, no z-scoring. 

The pairwise head with dot_out = x1 * x2 is the intended design and is stable when used with raw OE targets in the expected range. The reason your model diverges is most likely that the target distribution is different from what the model was designed for. When targets have the wrong distribution, MSE gradients from mismatched values push the representations to grow, and the elementwise product amplifies this quadratically, causing the collapse you observe.

The OneCycleLR scheduler is configured with steps_per_epoch=len(train_loader), so it expects to be called once per batch, which is what the code does. This is consistent and it is not the source of the instability.

You are correct that the models/ directory needs to exist beforehand. Thanks for flagging this. 

If you need Hi-C target labels, please let me know. We’re happy to share the data with you. 

I also wanted to mention that we have released an updated version, EPCOT v2. This version is trained on a larger and more diverse dataset and supports additional modalities. Here is the EPCOT v2 paper:https://academic.oup.com/nar/article/53/21/gkaf1269/8340987, and the GitHub repository:https://github.com/liu-bioinfo-lab/general_AI_model. It also includes COP task, in case you are interested in exploring it further.

**Comment by liushiji0511** [comment]
The genomic window used by EPCOT v2 is too short for my application. If I use EPCOT v2, I would like to retrain the model with a window size of approximately 1 Mb. Does EPCOT v2 conveniently support retraining with a larger genomic window?

________________________________
发件人: Linghua Jiang ***@***.***>
发送时间: 2026年4月20日 3:33
收件人: liu-bioinfo-lab/EPCOT ***@***.***>
抄送: liushiji0511 ***@***.***>; Author ***@***.***>
主题: Re: [liu-bioinfo-lab/EPCOT] Validation output diverges to very large constants during GM12878 Hi-C/OE fine-tuning, while training loss remains low (Issue #9)

[https://avatars.githubusercontent.com/u/77561062?s=20&v=4]linghuajiang left a comment (liu-bioinfo-lab/EPCOT#9)<https://github.com/liu-bioinfo-lab/EPCOT/issues/9#issuecomment-4276656666>

Hi, thank you for your interest in EPCOT. I think the main problem might be your target data preparation. In our original pipeline, Hi-C targets are raw OE values with no additional transformation C no log, no clipping, no z-scoring.

The pairwise head with dot_out = x1 * x2 is the intended design and is stable when used with raw OE targets in the expected range. The reason your model diverges is most likely that the target distribution is different from what the model was designed for. When targets have the wrong distribution, MSE gradients from mismatched values push the representations to grow, and the elementwise product amplifies this quadratically, causing the collapse you observe.

The OneCycleLR scheduler is configured with steps_per_epoch=len(train_loader), so it expects to be called once per batch, which is what the code does. This is consistent and it is not the source of the instability.

You are correct that the models/ directory needs to exist beforehand. Thanks for flagging this.

If you need Hi-C target labels, please let me know. We’re happy to share the data with you.

I also wanted to mention that we have released an updated version, EPCOT v2. This version is trained on a larger and more diverse dataset and supports additional modalities. Here is the EPCOT v2 paper:https://academic.oup.com/nar/article/53/21/gkaf1269/8340987, and the GitHub repository:https://github.com/liu-bioinfo-lab/general_AI_model. It also includes COP task, in case you are interested in exploring it further.

―

**Comment by linghuajiang** [comment]
EPCOT v2 is currently designed for a fixed 600kb input window. Due to the model architecture, particularly local encoder part, extending it to a longer window such as 1 Mb would not be straightforward.
